### ISSUE TITLE
Scalability tests failing on Jenkins due to non starting pods

### ISSUE THREAD

**Issue by wojtek-t** [GoogleCloudPlatform/kubernetes#10282]
There were a lot of failures in our scalability tests during last day (more than 50%). There seem to be two main reasons:
- timeouts while waiting for deleting the previous namespace - this should be fixed by #10225 
- some pods not starting

I started looking deeper into those non-starting pods. Here is an example that contributed to the failure:
- Jenkins scalability run 1539 - Density test
- we are looking only on pods from "e2e-tests-density-xyfx4" namespace
- look at the node "e2e-scalability-minion-k2n5"
- scheduler scheduled 10 pods onto that node
- it seems that at most 7 pods were actually started (7 logs "Need to restart pod infra container ...")
- then there is a log: "GET /healthz: (928.173627ms) 500"
- after that all the pods from that node were killed: "killing unwanted pod"

Links to logs:
- Jenkins: http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-scalability/1539/consoleFull
- kubelet logs: https://storage.googleapis.com/kubernetes-jenkins/logs/kubernetes-e2e-gce-scalability/1539/_artifacts/104.197.100.6%3A22-kubelet.log
- scheduler logs: https://storage.googleapis.com/kubernetes-jenkins/logs/kubernetes-e2e-gce-scalability/1539/_artifacts/104.154.56.51%3A22-kube-scheduler.log

To summarize, my hypothesis is that:
- Kubelet syncPods seems to hang in this example
- then I don't see any log saying that Kubelet was restarted, but it again starts to process thing
- however, due to the reason I don't understand it kills all the pods that were running there

I think that if the Kubelet didn't kill all the pods after again syncing pods, this wouldn't cause test failures. But anyway - there seem to be some problem here.

@dchen1107 @yujuhong - can you please explain what exactly is happening there? I think that this is similar to what was observed in #9895, but I don't understand why all the pods are being killed.

cc @fgrzadkowski @piosz @davidopp @bprashanth

**Comment by bprashanth** [comment]
The nodecontroller killed the pods:
```
I0623 19:14:12.023381       5 nodecontroller.go:388] Adding pods to evict2: 2015-06-23 19:14:12.023366352 +0000 UTC is later than 2015-06-23 19:14:12.02335977 +0000 UTC + 4m20s
I0623 19:14:12.109886       5 podevictor.go:110] PodEvictor is evicting Pods on Node: e2e-scalability-minion-k2n5
I0623 19:14:12.109932       5 nodecontroller.go:427] Delete all pods from e2e-scalability-minion-k2n5
```

You also see in the kubelet logs:
```
I0623 18:53:39.083844    3176 kubelet.go:1886] Recording NodeReady event message for node e2e-scalability-minion-k2n5
...

I0623 19:16:01.729467    3176 kubelet.go:1886] Recording NodeReady event message for node e2e-scalability-minion-k2n5
```
Which is too spaced out for node heartbeats.


As you noted there was a 500 on healthz, this might've led to monit restarting the kubelet though I don't see it in the logs. It seems like there were also memory issues on the node just before the 500 (note the du failure below):

```
I0623 19:15:59.168733    3176 server.go:608] POST /stats/container/: (6m5.701087699s) 0 [[Go 1.1 package http] 10.245.28.3:54251]
I0623 19:16:00.255319    3176 container.go:369] Failed to update stats for container "/3078d60a22bddaf3852ee52f7a9f1b36df184a3557fa4482abce167f4473bebc": du command failed on /var/lib/docker/aufs/diff/3078d60a22bddaf3852ee52f7a9f1b36df184a3557fa4482abce167f4473bebc with output  - fork/exec /usr/bin/du: cannot allocate memory, continuing to push stats
I0623 19:16:00.465821    3176 server.go:608] POST /stats/container/: (916.691646ms) 0 [[Go 1.1 package http] 10.245.43.25:34666]
E0623 19:16:00.693160    3176 reflector.go:183] watch of *api.Service ended with: 401: The event in requested index is outdated and cleared (the requested history has been cleared [48787/7621]) [49786]
I0623 19:16:00.750086    3176 server.go:608] GET /healthz: (928.173627ms) 500
goroutine 4066 [running]:
```

I see errors indicating SyncPods is killing containers while the podworker is creating pods (this is normal if you delete as the pods are coming up):
```
I0623 19:16:03.062023    3176 manager.go:1114] Killing container with id "1554c76fdb2e75b76a7ee49c89871825063b5d33a5e7ef38a174086a7e12f0ea"
...
E0623 19:16:04.047491    3176 manager.go:1533] 
Error running pod "density3000-04eebfdf-19db-11e5-bf18-42010af01555-vjjpb_e2e-tests-density-xyfx4" container "density3000-04eebfdf-19db-11e5-bf18-42010af01555": API error (500): Cannot start container 63f1044d574ce513180ae6cb7f6e2c3bbe905145d02b0331fce4df215cb6690a: cannot join network of a non running container: 1554c76fdb2e75b76a7ee49c89871825063b5d33a5e7ef38a174086a7e12f0ea
```

The pod in the example (density3000-04eebfdf-19db-11e5-bf18-42010af01555-vjjpb) had this happen:
- scheduler assigned (19:08:53)
- Kubelet picked up (19:08:54)
- Node controller deletes (19:14:14)
- suspicious 500 (19:16)
- kuelet restarts mentioned above (19:16)

**Comment by davidopp** [comment]
Just for my edification, what are the circumstances under which pods can be killed?
* explicit user request
* kubelet failing healthcheck for longer than the node controller threshold (5 min or whatever it is)
* any others? i thought some other issue implied that restarting kubelet kills pods -- but that is not true, is it?

**Comment by bprashanth** [comment]
> i thought some other issue implied that restarting kubelet kills pods -- but that is not true, is it?

In theory it shouldn't, we should skip deletion till all our sources are ready (need to convince myself that there isn't a race that could lead to container restarts here).

>any others?

Containers die with docker restarts and OOMKill. The kubelet won't meddle with the pods in the apiserver directly, if it ends up killing some by mistake it'll notice the next time it syncs and start them.

**Comment by wojtek-t** [comment]
I didn't observe any such failure in last few days. Are we just lucky or was it fixed? If so - can someone point me to the PR fixing that?

**Comment by bprashanth** [comment]
This will still happen if a node goes into the NotReady state at the wrong time. The most common case of this occurrence 
 is a cluster addon hogging resources. Setting limits should help https://github.com/GoogleCloudPlatform/kubernetes/pull/10260, but I think that pr got reverted.

**Comment by wojtek-t** [comment]
@bprashanth Is there a dedicated issue for it? If not, maybe we should rename this one (i.e. scalability tests are no longer failing because of that particular reason).

**Comment by wojtek-t** [comment]
Yes - this is still an issue. The failure with exactly the same logs just happened tonight:
http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-scalability/1649/consoleFull

For the reference:
- the node that had problems: e2e-scalability-minion-bv49 (104.154.79.113-22-kubelet.log in GCS)
- again there are logs like this:

        2814 I0701 05:44:17.407515   18374 container.go:369] Failed to update stats for container "/2e20533074fb351801e684c5393474f9005f4ba8b450bb664c421e5837d19216": du command failed on /var/lib/do     cker/aufs/diff/2e20533074fb351801e684c5393474f9005f4ba8b450bb664c421e5837d19216 with output  - fork/exec /usr/bin/du: cannot allocate memory, continuing to push stats

**Comment by wojtek-t** [comment]
This just happened again.

@bprashanth @lavalamp @davidopp 
do you think that we should be reporting it as an error in scalability tests? If the Kubelet is dead and node controller is deleting the pods running on that node, ReplicationController will recreate them and they will be scheduled somewhere else. So maybe we should just log it and wait until all of them are running instead of failing the whole test in that situation? Thoughts?

**Comment by wojtek-t** [comment]
cc @dchen1107 
on the other hand, setting limits for addons may also solve this issue

**Comment by bprashanth** [comment]
The feeling I got from https://github.com/GoogleCloudPlatform/kubernetes/issues/9788 was, when something starts hogging memory (eg: heapster without cgroup limits) we wait for it to get oom killed. We increase the chances that the oom killer kills the right things by setting oom score. In theory, this should ensure that we never kill a system daemon and the node should not flip into NotReady. 

If we somehow end up using 95% memory without hitting the oom limit for 5m, the kubelet will be too slow to update its status and the node controller will kill the pods. The most likely reason to hit this situation is a misbehaving cluster addon. There are other reasons the kubelet might not have updated status in 5m though, eg: goroutine leak leading to cpu starvation, monit restart loop etc. I think the way to handle this is by detecting NotReady up front and dumping information about what was on the node just before it went down (https://github.com/GoogleCloudPlatform/kubernetes/pull/10329, https://github.com/GoogleCloudPlatform/kubernetes/issues/10306).

We could do this in an independent soak test, but I think it indicates system instability.

**Comment by lavalamp** [comment]
@dchen1107 this is your hobby horse, I think.

@wojtek-t I'm guessing that setting limits for addons will resolve this, so let's wait for that and see?

**Comment by wojtek-t** [comment]
@lavalamp 

>
@wojtek-t I'm guessing that setting limits for addons will resolve this, so let's wait for that and see?

I'm fine with waiting - those failures are relatively rare so it's not a big problem to look into them from time to time to check whether it's again the same problem or a new regression.

**Comment by wojtek-t** [comment]
It seems that setting limits for addongs - #10653 fixed the main issue here (there wasn't such failure since merging that PR).

Currently the only failures are are because of too high latency metrics.

**Comment by davidopp** [comment]
Great -- please close the issue unless you think there is something left to do here. (Other than investigate the latency failures, which is unrelated to this issue.)
